File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.java
class Solution {
    public int minRefuelStops(int target, int s, int[][] stations) {
        int n = stations.length;
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        int ans = 0;
        int cur = s;
        for(int i = 0 ; i <= n ; i++){
            while ((i < n && cur < stations[i][0]) || (i == n && cur < target)) {
                if (pq.isEmpty()) return -1;
                cur -= pq.poll();
                ans++;
            }
            if (i != n) pq.offer(-stations[i][1]);
        }
        return ans;

    }
}

File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.cpp
class Solution {
public:
    int minRefuelStops(int tar, int s, vector<vector<int>>& stations) {
        priority_queue<int > pq;
        int n = stations.size();
        int cur = s;
        int cnt = 0;
        stations.push_back({tar , 0});
        for(int i = 0 ; i <= n ; i++){
            while( cur < stations[i][0]){
                if(pq.empty()) return -1;
                cur += pq.top();
                pq.pop();
                cnt++;
            }
            pq.push(stations[i][1]);
        }

        return cnt;
    }
};